libcaf_net/caf/net/datagram_transport.hpp

```diff
@@ -219,75 +144,66 @@ class datagram_transport {
         bytes.emplace_back(std::move(*buf));
       }
     }
+
+    std::vector<std::vector<byte>*> get_buffer_ptrs() {
+      std::vector<std::vector<byte>*> ptrs;
+      for (auto& buf : bytes)
+        ptrs.emplace_back(&buf);
+      return ptrs;
+    }
   };
 
 private:
   // -- utility functions ------------------------------------------------------
 
-  static buffer_type next_buffer_impl(buffer_cache_type cache) {
-    if (cache.empty()) {
-      return {};
-    }
-    auto buf = std::move(cache.back());
-    cache.pop_back();
-    return buf;
+  void prepare_next_read() {
+    this->read_buf_.resize(max_datagram_size);
   }
 
   bool write_some() {
-    CAF_LOG_TRACE(CAF_ARG(handle_.id));
+    CAF_LOG_TRACE(CAF_ARG2("handle", this->handle_.id));
     // Helper function to sort empty buffers back into the right caches.
     auto recycle = [&]() {
       auto& front = packet_queue_.front();
       auto& bufs = front.bytes;
       auto it = bufs.begin();
-      if (header_bufs_.size() < header_bufs_.capacity()) {
+      if (this->header_bufs_.size() < this->header_bufs_.capacity()) {
         it->clear();
-        header_bufs_.emplace_back(std::move(*it++));
+        this->header_bufs_.emplace_back(std::move(*it++));
       }
-      for (;
-           it != bufs.end() && payload_bufs_.size() < payload_bufs_.capacity();
+      for (; it != bufs.end()
+             && this->payload_bufs_.size() < this->payload_bufs_.capacity();
            ++it) {
         it->clear();
-        payload_bufs_.emplace_back(std::move(*it));
+        this->payload_bufs_.emplace_back(std::move(*it));
       }
       packet_queue_.pop_front();
     };
     // Write as many bytes as possible.
     while (!packet_queue_.empty()) {
       auto& packet = packet_queue_.front();
-      std::vector<std::vector<byte>*> ptrs;
-      for (auto& buf : packet.bytes)
-        ptrs.emplace_back(&buf);
-      auto write_ret = write(handle_, make_span(ptrs), packet.id);
+      auto ptrs = packet.get_buffer_ptrs();
+      auto write_ret = write(this->handle_, make_span(ptrs), packet.id);
       if (auto num_bytes = get_if<size_t>(&write_ret)) {
-        CAF_LOG_DEBUG(CAF_ARG(handle_.id) << CAF_ARG(*num_bytes));
+        CAF_LOG_DEBUG(CAF_ARG(this->handle_.id) << CAF_ARG(*num_bytes));
         CAF_LOG_WARNING_IF(*num_bytes < packet.size,
                            "packet was not sent completely");
         recycle();
       } else {
         auto err = get<sec>(write_ret);
         if (err != sec::unavailable_or_would_block) {
-          CAF_LOG_DEBUG("send failed" << CAF_ARG(err));
-          dispatcher_.handle_error(err);
+          CAF_LOG_ERROR("write failed" << CAF_ARG(err));
+          this->next_layer_.handle_error(err);
           return false;
         }
+        CAF_LOG_DEBUG("write returned `unavailable_or_would_block`");
```

Comment:
Does this fit our usually style for return values?

Reply:
Not sure what you mean?

Reply:
I was talking about the backticks around the error. If we write the complete message ourselves we could just say what is happening instead of printing the name of the variable we use internally for this error code.